iot-e2e-tests/common/src/test/java/tests/integration/com/microsoft/azure/sdk/iot/iothub/twin/TwinPnPTests.java

```diff
@@ -0,0 +1,185 @@
+import com.microsoft.azure.sdk.iot.service.exceptions.IotHubException;
+import com.microsoft.azure.sdk.iot.service.Device;
+import com.microsoft.azure.sdk.iot.service.Module;
+import org.junit.After;
+import org.junit.AfterClass;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+import tests.integration.com.microsoft.azure.sdk.iot.helpers.DeviceConnectionString;
+import tests.integration.com.microsoft.azure.sdk.iot.helpers.IntegrationTest;
+import tests.integration.com.microsoft.azure.sdk.iot.helpers.TestConstants;
+import tests.integration.com.microsoft.azure.sdk.iot.helpers.Tools;
+import tests.integration.com.microsoft.azure.sdk.iot.helpers.annotations.IotHubTest;
+import tests.integration.com.microsoft.azure.sdk.iot.helpers.annotations.StandardTierHubOnlyTest;
+
+import java.io.IOException;
+import java.net.URISyntaxException;
+import java.util.*;
+
+import static com.microsoft.azure.sdk.iot.device.IotHubClientProtocol.*;
+import static com.microsoft.azure.sdk.iot.device.IotHubClientProtocol.MQTT_WS;
+import static org.junit.Assert.assertEquals;
+
+/**
+ * Test class containing all tests to be run for PnP.
+ */
+@IotHubTest
+@RunWith(Parameterized.class)
+public class TwinPnPTests extends IntegrationTest
+{
+    protected static String iotHubConnectionString = "";
+    private static RegistryManager registryManager;
+    private static final String ModelId = "dtmi:com:test:e2e;1";
+
+    @Parameterized.Parameters(name = "{0}")
+    public static Collection inputs() throws IOException
+    {
+        iotHubConnectionString = Tools.retrieveEnvironmentVariableValue(TestConstants.IOT_HUB_CONNECTION_STRING_ENV_VAR_NAME);
+        isBasicTierHub = Boolean.parseBoolean(Tools.retrieveEnvironmentVariableValue(TestConstants.IS_BASIC_TIER_HUB_ENV_VAR_NAME));
+        isPullRequest = Boolean.parseBoolean(Tools.retrieveEnvironmentVariableValue(TestConstants.IS_PULL_REQUEST));
+
+        registryManager = RegistryManager.createFromConnectionString(iotHubConnectionString);
+
+        List inputs =  Arrays.asList(
+                new Object[][]
+                        {
+                                {MQTT},
+                                {MQTT_WS},
+                        }
+        );
+
+        return inputs;
+    }
+
+    private static class TestDevice
+    {
+        String deviceId;
+        String moduleId;
+        DeviceClient deviceClient;
+        ModuleClient moduleClient;
+    }
+
+    public TwinPnPTests.TwinPnPTestInstance testInstance;
+
+    public TwinPnPTests(IotHubClientProtocol protocol) throws IOException
+    {
+        this.testInstance = new TwinPnPTestInstance(protocol);
+    }
+
+    public class TwinPnPTestInstance
+    {
+        public IotHubClientProtocol protocol;
+        private Device deviceForRegistryManager;
+        private Module moduleForRegistryManager;
+
+        private DeviceTwin sCDeviceTwin;
```

Comment:
I know you are following the pattern here, but I'd rather this name specified that it is a client. Something like "twinServiceClient" or "twinClient" instead of "sCDeviceTwin"

Reply:
I like that suggestion too. I would follow the same pattern like other tests for less confusion though.

Reply:
Now would be a good time to rename the other uses of this naming then